eliot/_message.py

```diff
@@ -106,16 +125,49 @@ def write(self, logger=None, action=None):
         """
         if logger is None:
             logger = _output._DEFAULT_LOGGER
-        if action is None:
-            action = currentAction()
-        if action is None:
-            action = _defaultAction
-        contents = self._contents.copy()
-        contents["timestamp"] = self._timestamp()
-        if action is not None:
-            contents["task_uuid"] = action._identification["task_uuid"]
-            contents["task_level"] = action._nextTaskLevel().level
-        logger.write(contents, self._serializer)
+        logged = self._freeze(action=action)
+        logger.write(logged.asDict(), self._serializer)
+        return logged
+
+
+
+class LoggedMessage(namedtuple('LoggedMessage', (
```

Comment:
FWIW, I copied the code around me. Do you have any in-principle objections to a patch that just moves everything that's already immutable to `pyrsistent`?